### ISSUE TITLE
A "pure" version of inject

### ISSUE THREAD

**Issue by luisherranz** [mobxjs/mobx-react#454]
I've posted this on gitter, but I guess it makes more sense to open an issue here.

----

I want to know if is there any reason why an `inject` like this:
```js
inject(({ connection }, { id }) => ({
  isSelected: connection.selectedItem.id === id,
}))(NavItem)
```
couldn't trigger a re-render only when `isSelected` changes, and not every time `selectedItem` does.

The current (non-pure) behaviour is great for `observer` but when using `inject`, it could shallow compare the returned object. If props don't change, you don't need to re-render, even if the `inject` had to run again because it was invalidated by some observable

That way we wouldn't have to check it afterwards using `shouldComponentUpdate` or `PureComponent`. We always use something like this after `inject` (using `recompose`):
```js
compose(
  inject(({ connection }, { id }) => ({
      isSelected: connection.selectedItem.id === id,
  }),
  pure,
)(NavItem);
```

If for some reason that's not the desirable default behaviour, maybe something like `pureInject` could be added to `mobx-react`, which would extend from `PureComponent` instead of `Component`.

Any thoughts?

**Comment by luisherranz** [comment]
> Injector component doesn't have an access to wrapped component props

By my understanding of the code, not only it does have access to *wrapped component props*, but actually, the `shouldComponentUpdate` inside `observer` makes the wrapped component *pure*:

https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L259-L274

```js
// update if props are shallowly not equal, inspired by PureRenderMixin
return !shallowEqual(this.props, nextProps)
```

> The subject of inject is observer most of the time, so there would be a redundant component with a redundant shouldComponentUpdate check

When using `observer` if an observable changes, there's no way to know if the re-render was necessary or not. I agree. But when using `inject`, if the props returned by `inject` didn't change, you can most certainly assure that the re-render was not necessary.

> Also implementing this in user land should be trivial

Sure, it's trivial to implement. But that doesn't mean it couldn't be included in the library.

To sum up, my question is: why shallow comparing the *wrapped component props* is the desired behaviour and shallow comparing the *injected props* is not.

**Comment by urugator** [comment]
`observer` patches the component, but `inject` wraps the component into `Injector` component which obviously doesn't have an access to props of it's children (wrapped component)...

**Comment by urugator** [comment]
> But when using inject, if the props returned by inject didn't change, you can most certainly assure that the re-render was not necessary.

What you're saying is: Whenever `inject` is used, the component is certainly pure ... how did you get to this conclusion?
Mobx makes observers pure by default, because it presumes that you're using observables in them.

**Comment by luisherranz** [comment]
> What you're saying is: Whenever inject is used, the component is certainly pure ... how did you get to this conclusion?

I came to that conclusion because `inject` is using `observer`:
https://github.com/mobxjs/mobx-react/blob/master/src/inject.js#L118
and `observer` attaches a `shouldComponentUpdate` with shallow compare of props passed to `Injector` (wich are then passed down to the wrapped component):
https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L259-L274

> Mobx makes observers pure by default, because it presumes that you're using observables in them.

Yeah, right?

I'll try to explain my point a bit further.

You can use `inject` in a similar fashion to `observer`:
```js
let NavItem = observer(({ isSelected }) => isSelected ? <div>Selected!</div> : null);

NavItem = inject(({ connection }, { id }) => ({
  isSelected: connection.selectedItem.id === id,
}))(NavItem)
```
versus
```js
const NavItem = observer(({ id }) => {
   const isSelected = connection.selectedItem.id === id;
   return isSelected ? <div>Selected!</div> : null;
});
```

If you want to avoid unnecessary re-renders when using `observer`, you'd have to create computed values in your store, and pass them directly:

```js
const NavItem = observer(({ id }) => connection.isSelected(id) ? <div>Selected!</div> : null);
```

But maybe you don't want to flood your store with "one-time-use" computed values and probably that logic belongs to the specific component more than to the general store.

So, why don't we give the opportunity to use `inject` to create this type of computed values, specific to the component and in an *automatic way*, with the very same code used before?
 ```js
inject(({ connection }, { id }) => ({
  isSelected: connection.selectedItem.id === id,
}))(NavItem)
```
All we'd have to do is shallow compare the returned object.

I might be wrong, but my point is: `inject` should not trigger a re-render when the returned props don't change.

**Comment by urugator** [comment]
The injector can be invoked (on prop/observable change), can return the (shallowly) same set of props , but the component still may need to be updated if it's not designed as pure (and inject doesn't know how the component is designed):
Consider component like: `const NavItem = ({ item }) => item.text;`
Such component is designed in a way that it should be updated whenever it recieves new props from parent (react default behavior).
If you wrap it to `inject`, to supply `item` or whatever, nothing changes in this regard. Even if the `item` stays the same, the `item.text` could have been changed, so the component has to re-render.

I think I undestand your (valid) usage, I am just suggesting that it's not always the case, therefore the conclusion is wrong...

**Comment by luisherranz** [comment]
> Consider component like: const NavItem = ({ item }) => item.text;

Ok, I get that now. Great example.

> I think I undestand your (valid) usage, I am just suggesting that it's not always the case, therefore the conclusion is wrong...

My initial question was about the reasons that was not the default behaviour. Now that that is solved (thanks about that), I still wonder if adding to this library something, easy to use and well documented, would help people solve this unnecessary re-renders without having to rely on libraries like `recompose`, or switching to classes just to extend `PureComponent` or add `shouldComponentUpdate`'s.
